Optionally use a custom libFuzzer.a archive

deweerdt opened the pull request:
Tweak build.rs so that it's possible to use a custom libFuzzer.a
archive. This is necessary when using OSS-Fuzz
(https://github.com/google/oss-fuzz): the build
system has to link with the libFuzzer provided by
the `LIB_FUZZING_ENGINE` environment variable (see
https://github.com/google/oss-fuzz/blob/master/docs/new_project_guide.md)

frewsxcv commented:
oh wow, this is great! i've been meaning to get this working on libfuzzer. these changes look good to me. @nagisa @Manishearth any thoughts?

Manishearth commented:
r? @nagisa 

overall lgtm

nagisa commented on build.rs:
```diff
@@ -1,17 +1,30 @@
+        let custom_lib_dir = ::std::path::PathBuf::from(&custom);
+        let custom_lib_dir = custom_lib_dir.parent().unwrap().to_string_lossy();
+
+        let custom_lib_name = ::std::path::PathBuf::from(&custom);
```
It is unnecessary to create two `PathBuf`s here.

nagisa commented on build.rs:
```diff
@@ -1,17 +1,30 @@
+
+        let custom_lib_name = ::std::path::PathBuf::from(&custom);
+        let custom_lib_name = custom_lib_name.file_stem().unwrap().to_string_lossy();
+        let custom_lib_name = &custom_lib_name[3..];
```
You cannot assume filename begins with `lib`. Use `str::trim_left_matches` instead.

nagisa commented on build.rs:
```diff
@@ -1,17 +1,30 @@
+
+        println!("cargo:rustc-link-search=native={}", custom_lib_dir);
+        println!("cargo:rustc-link-lib=static={}", custom_lib_name);
+        println!("cargo:rustc-link-lib=stdc++");
```
Seems fairly not-okay to assume all the custom fuzzers will be written in C++. It is fine to leave as-is for now, but should at least have a `FIXME` comment on top of this.

nagisa commented:
Oh, also, I’ve experimented a number of times with upgrading our libFuzzer before but the never versions of the fuzzer simply do not work. I have a variety reasons in mind why this could happen but never bothered investigating it further. Seems fairly plausible that anybody using this feature (even oss-fuzz) could hit the same issue.

deweerdt commented:
Thank you for your review, @nagisa. I believe that 9ba1905 addresses all your comments so far.

> I hope it is pretty obvious that using CUSTOM_LIBFUZZER_PATH will come with no stability guarantees or support whatsoever.

It's clear to me.

> Oh, also, I’ve experimented a number of times with upgrading our libFuzzer before but the newer versions of the fuzzer simply do not work.

I've managed to use the current tip of libFuzzer, but for that I had to patch `cargo-fuzz` with this:

```

nagisa commented:
@deweerdt thanks!

You want to send patch to cargo-fuzz as well?

bors r+

deweerdt commented:
> You want to send patch to cargo-fuzz as well https://github.com/rust-fuzz/cargo-fuzz/pull/118
